src/bindings/python/flux/resource/ResourceSet.py

```diff
@@ -0,0 +1,164 @@
+###############################################################
+# Copyright 2020 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+###############################################################
+
+import json
+
+from flux.idset import IDset
+from flux.resource import Rlist
+from flux.resource.ResourceSetImplementation import ResourceSetImplementation
+
+
+class ResourceSet:
+    def __init__(self, rset=None, impl=None, version=1):
+        """
+        ResourceSet object constructor.
+
+        :param rset: an R object serialized as JSON string or dict
+        :param version:  R specification version
+        :param impl: A resource set "implementation" object conforming
+                     to the ResourceSetImplementation interface
+        :raises ValueError: If `impl` not a valid ResourceSet implementation
+                            or version not supported
+        """
+        self._state = None
+        if impl is not None:
+            if not isinstance(impl, ResourceSetImplementation):
+                tstr = type(impl)
+                raise ValueError(f"{tstr} is not a ResourceSet implementation")
+            self.impl = impl
+            self.version = impl.version
+            return
+
+        if rset is not None:
+            rdict = rset
+            if isinstance(rdict, str):
+                rdict = json.loads(rdict)
+            version = rdict["version"]
```

Comment:
b/c of the `return` in the first branch here, shouldn't the `rset is not None` branch be listed first?  i.e. if user sets rset != NONE and impl != NONE, rset won't be set.

Reply:
oh wait, i might have been confused about how this object creation works.  i guess its really about initializing `rset` or `impl`, but not both.  Perhaps just a comment should be done that `impl` overrides `rset` or something like that.

Reply:
Yes, comments are a good idea. I actually didn't have an easy time making a ResourceSet that can be initialized from an existing implementation handle, or a valid _R_ string (currently version 1 only supported), or a dict representation of an r string.

If an implementation handle is passed in, then there is no need for a resource set object from which to initialize the implementation.

Reply:
Oh, also it doesn't matter if rset is `None`, this just instantiates an empty `ResourceSet`

Reply:
although my gut feeling says this is a bad design pattern, would a:

```
if isinstance (ResourceSetImplementation):
  ...
else if isinstance (str):
  ...
else if isinstance (dict);
  ...
```

work?

Reply:
I actually kind of like that, and  the `IDset` and `Hostlist` classes kind of use that approach, so I'm not sure why I didn't think of it here. (I think it was because I first had the alternate constructor approach)

Reply:
Yeah, this is much better and the calling ergonomics are greatly improved. Thanks @chu11!!
```python
class ResourceSet:
    def __init__(self, arg=None, version=1):
        """
        ResourceSet object constructor.

        :param arg: Argument from which to construct a ResourceSet. `arg`
                    may be a serialized R string, a decoded Mapping of
                    an R string, or a valid ResourceSet implementation
                    (an instance of ResourceSetImplementation)

        :param version: R specification version

        :raises TypeError: A ResourceSet cannot be instantiated from arg
        :raises ValueError: Invalid R version, or invalid R encoding

        All parameters are optional. ResourceSet() will initialize an
        empty, version 1 ResourceSet object.
        """
        self._state = None

        if isinstance(arg, ResourceSetImplementation):
            #  If argument is a resource set implementation, instantiate
            #   from that and return immediately
            self.impl = arg
            self.version = arg.version
            return

        if isinstance(arg, str):
            #  If arg is a string, assume an encoded R representation.
            #  decode to a mapping:
            arg = json.loads(arg)

        if isinstance(arg, Mapping):
            #  If argument is a mapping, grab version field for later use
            version = arg["version"]
        elif arg is not None:
            # arg must be ResourceSetImplementation, Mapping, string or None:
            tstr = type(arg)
            raise TypeError(f"ResourceSet cannot be instantiated from {tstr}")

        #  Instantiate implementation from version
        #  note: only version 1 supported for now
        if version == 1:
            self.version = 1
            self.impl = Rlist(arg)
        else:
            raise ValueError(f"R version {version} not supported")

```

Reply:
I like it.

Only comment is I think `KeyError` can also happen if they pass in a bad dict?

and don't know if result from `json.loads()` needs to be checked if the user passed in something ok? (not sure if that raises an error on bad input or not).

Reply:
Yeah, json will throw JsonDecodeError or something. `Rlist` constructor (or other classes in the future) will throw its own exceptions as well. I can document the `KeyError`, but is it really common practice to document every exception that might be thrown from a function (might as well list them all)?

Reply:
> is it really common practice to document every exception that might be thrown from a function (might as well list them all)?

Good point, its probably not necessary.  The `JsonDecodeError` one would hopefully be obvious if a user saw the stacktrace.  But not sure if the KeyError one would be.

Reply:
Yeah, I documented that and the JsonDecodeError, and added tests. Thanks!